### ISSUE TITLE
Update Open MPI to require a C11 compiler

### ISSUE THREAD

**Issue by hjelmn** [open-mpi/ompi#3879]
We are discussing updating the compiler requirement for Open MPI to require C11. The features we want to require are:

 - C11 atomics. These will completely replace the hand-coded assembly in Open MPI.

 - Thread local support. This feature will completely replace the thread local storage code provided by opal.

 - aligned_alloc. Right now we use a memalign, posix_memalign, or malloc + hand aligning. This will bump the minimum glibc up to 2.16. If 2.16 is too new we can wrap this function in opal/util.

Like to haves:

 - Static assert

 - Generics

Compiler status pages:

 - Intel: https://software.intel.com/en-us/articles/c11-support-in-intel-c-compiler . Intel is missing _Atomic even in the latest versions. We will have to use atomic_* to support this compiler.

 - GCC: https://gcc.gnu.org/wiki/C11Status

 - Clang: v3.1+ started adding support in v3.1. C11 support was enabled by default in 2014.

 - PGI: Supported since at least late 2014. Will have to verify older release.

**Comment by hjelmn** [comment]
One comment. LANL tends to support some old compiler versions to support fortran. This can be handled by using a supported compiler and using the older fortran compiler for the bindings. Example: ```./configure CC=gcc-6 FC=ifort```.

**Comment by hjelmn** [comment]
One question is whether we should continue to offer ```opal_atomic_*``` functions or just go straight to using ```atomic_*``` everywhere. There are benefits to both but the later is a little more work and is probably cleaner in the long run.

**Comment by hjelmn** [comment]
Basic configury is complete. Will push the branch to my repository shortly.

```
configure: checking if gcc-7 supports C11 without a flag
checking if gcc-7 supports C11 thread local storage... yes
checking if gcc-7 supports C11 atomic variables... yes
checking if gcc-7 supports C11 _Atomic keyword... yes
checking if gcc-7 supports C11 _Generic keyword... yes
configure: no flag required for C11 support
```

**Comment by hjelmn** [comment]
We may want to keep some of the inline assembly going forward. It looks like gcc v7 removed support for lock-free compare-and-set for 128-bit integers on x86-64. This was done because the load/store of these integers is not atomic unless CAS is used. We do not care about this in the lifo/fifo implementations. A similar problem affects the LL/SC fifo/lifo implementations. There is no C11 atomic support for these atomics.

For now my C11 ompi prototype will leave the x86-64 CAS128 and PPC/ARM64 LL/SC atomics and remove the rest.

**Comment by hjelmn** [comment]
Another note about C11 atomics. The language guarantees the atomicity of arithmetic and bit-wise operators (+, -, *, ^, etc). So these are equivalent:

```
atomic_int foo;

atomic_fetch_add (&foo, 1);
```

```
atomic_int foo;

foo += 1;
```

The benefit of the ```atomic_*``` calls is the ability to add _explicit and change the memory model.

This behavior has an impact on the ```OPAL_THREAD_*``` macros. Those will have the cast away the _Atomic to when ```opal_using_threads() == false``` to ensure that an atomic is not emitted by the compiler. Example:

```c
int32_t opal_thread_add_32 (opal_atomic_int32_t *ptr, int32_t value) {
    if (opal_using_threads ()) {
        return (*p += value);
    }

    return (*((int32_t *) p) += value);
}
```

**Comment by bosilca** [comment]
This will prevent us from declaring the members of the structures as atomic, as not all operations on these members need to be atomic.

**Comment by hjelmn** [comment]
To some degree yes. We have to be careful and always cast away the atomic where needed. I am working through out current atomic usage in opal now. Will have a branch up next week.

**Comment by hjelmn** [comment]
Well, thats fun. We will not be able to use arithmetic operations directly on atomic types due to icc. We will have to use the ```atomic_*``` API.

```c
#include <stdlib.h>
#include <stdatomic.h>

atomic_int foo;

int main () {
  foo += 1;

  return foo;
}
```

```
foo.c(7): error: expression must have arithmetic or pointer type
    foo += 1;
    ^

foo.c(9): error: return value type does not match the function type
    return foo;
           ^

compilation aborted for foo.c (code 2)
```

**Comment by hjelmn** [comment]
I am almost leaning toward requiring Open MPI to be built with a truly compliant C11 compiler. This would mean no icc. This does not prevent us from building a version of Open MPI for a the intel compiler suite. See #4088.

**Comment by bwbarrett** [comment]
I would strongly push back on this.  Users have had it drilled into them for years that they should build Open MPI with the compiler they build their apps with.  I think changing that now will cause significant heartache.  While there's developer efficiencies in using C11 atomics, there's no technical reason that we *must* use them.

**Comment by hjelmn** [comment]
@bwbarrett Yeah, we may have to drop the whole idea of using C11 atomics until intel catches up if we are uncomfortable with this. I suggest this because C compilers are generally link compatible with each other and it might be worth looking at.

If not we can still look at _Thread_local and a couple of the other C11 features that Intel does support correctly.

**Comment by hjelmn** [comment]
So here is the C11 matrix so far:

|Compiler|First Version|C11 Atomics|_Thread_local|_Generic|_Static_assert|
|---------|-------------|------------|-------------|---------|--------------|
|Intel|16|No|Yes|Yes|Yes|
|gcc|4.9.0|Yes|Yes|Yes|Yes|
|clang|3.5|Yes|Yes|Yes|Yes|
|xlc|12.1|Yes (no stdatomic.h)|Yes|Yes|Yes|
|pgi|

**Comment by hjelmn** [comment]
@bwbarrett Just want to point out that may be some technical reasons to use the C11 atomics. There may be performance advantages to using compiler atomics (C11, __atomic builtins) because the compiler has more information. Not saying they are faster, just that they could be. That is part of the reason I wanted to do this investigation.

**Comment by hjelmn** [comment]
Looks like intel has finally added C11 atomic support. icc 18+ will be required when we finally require c11 atomic support. Depends on how long we want to support icc 16 and 17.

**Comment by bosilca** [comment]
Not for all types. Their support unfortunately only applies to 32 and 64 bits types.

**Comment by hjelmn** [comment]
@bosilca Shouldn't that be sufficient for our case? We only define 32 and 64-bit atomics. There is the 128-bit but we can define that specially since it is only used for ABA avoidance in the fifo and lifo.

**Comment by bosilca** [comment]
We can always do with 32 and 64 bits, for as long as we protect the 128 bits case. We'll lose some performance, but it will work. I just mention it here because their support is still partial.

**Comment by hjelmn** [comment]
@bosilca Yeah, I will note that in the table. Not sure why they have been dragging their feet on C11 atomics. C++ needs a similar backend and has similar semantics.

**Comment by hjelmn** [comment]
@bosilca Yeah, I will note that in the table. Not sure why they have been dragging their feet on C11 atomics. C++ needs a similar backend and has similar semantics.

**Comment by gpaulsen** [comment]
@awlauria @hjelmn Can we now close this?

**Comment by bwbarrett** [comment]
@gpaulsen I think we should close this issue.  I think we've pushed C11 usage as far as we sanely can at this point, and it's waiting for the world to catch up before we can take the next step.  There's a particularly ugly problem with ARM and POWER and LL/SC that we'd have to sort out and unless we can (which seems dubious), we're always going to have about the level of atomics complexity we have today.  The next major savings in complexity is probably dropping the gcc builtin code, which is a minor complexity advantage.
